File: langs/ruby/lib/dotlyte/workspace.rb
# frozen_string_literal: true

require "json"

module Dotlyte
  # Workspace support for monorepo environments.
  #
  # Detects monorepo roots (pnpm, turbo, nx, lerna, npm/yarn workspaces),
  # loads shared environment files, and merges per-package configs.
  #
  # @example
  #   info = Dotlyte::Workspace.find_monorepo_root
  #   configs = Dotlyte::Workspace.load_workspace(
  #     root: info.root,
  #     packages: %w[packages/api packages/web],
  #     shared_env_file: ".env"
  #   )
  module Workspace
    # Information about a detected monorepo.
    MonorepoInfo = Struct.new(:root, :type, :packages, keyword_init: true)

    # Marker files and their monorepo type.
    MARKERS = {
      "pnpm-workspace.yaml" => :pnpm,
      "turbo.json"          => :turbo,
      "nx.json"             => :nx,
      "lerna.json"          => :lerna
    }.freeze

    # Detect the monorepo root by walking up from the given directory.
    #
    # @param cwd [String] starting directory (default: Dir.pwd)
    # @return [MonorepoInfo, nil] info about the monorepo or nil if not found
    def self.find_monorepo_root(cwd: Dir.pwd)
      dir = File.expand_path(cwd)

      loop do
        # Check explicit monorepo markers
        MARKERS.each do |marker, type|
          marker_path = File.join(dir, marker)
          if File.exist?(marker_path)
            packages = detect_packages(dir, type, marker_path)
            return MonorepoInfo.new(root: dir, type: type, packages: packages)
          end
        end

        # Check package.json for workspaces field (npm/yarn)
        pkg_json = File.join(dir, "package.json")
        if File.exist?(pkg_json)
          begin
            pkg = JSON.parse(File.read(pkg_json))
            if pkg.key?("workspaces")
              packages = resolve_workspace_globs(dir, pkg["workspaces"])
              return MonorepoInfo.new(root: dir, type: :npm, packages: packages)
            end
          rescue JSON::ParserError
            # ignore malformed package.json
          end
        end

        parent = File.dirname(dir)
        break if parent == dir # filesystem root

        dir = parent
      end

      nil
    end

    # Load workspace configuration for a monorepo.
    #
    # @param root [String, nil] monorepo root directory (auto-detected if nil)
    # @param packages [Array<String>, nil] package directories to load
    # @param shared_env_file [String, nil] shared .env file name relative to root
    # @param prefix [String, nil] environment variable prefix to strip
    # @param env [String, nil] environment name
    # @return [Hash{String => Dotlyte::Config}] package name => Config mapping
    def self.load_workspace(root: nil, packages: nil, shared_env_file: nil, prefix: nil, env: nil)
      info = root ? MonorepoInfo.new(root: root, type: :unknown, packages: packages || []) : find_monorepo_root
      raise Error, "Could not detect monorepo root. Specify root: explicitly." unless info

      actual_root = info.root
      package_dirs = packages || info.packages || []
      shared = get_shared_env(actual_root, prefix: prefix)

      results = {}

      package_dirs.each do |pkg_dir|
        full_path = File.expand_path(pkg_dir, actual_root)
        next unless File.directory?(full_path)

        pkg_name = File.basename(pkg_dir)

        config = Dotlyte.load(
          cwd: full_path,
          defaults: shared,
          prefix: prefix,
          env: env
        )

        results[pkg_name] = config
      end

      results
    end

    # Load shared environment variables from the monorepo root.
    #
    # @param root [String] monorepo root directory
    # @param prefix [String, nil] env var prefix to strip
    # @return [Hash] shared environment data
    def self.get_shared_env(root, prefix: nil)
      env_file = File.join(root, ".env")
      return {} unless File.exist?(env_file)

      data = {}
      File.readlines(env_file, chomp: true).each do |line|
        line = line.strip
        next if line.empty? || line.start_with?("#")

        key, _, value = line.partition("=")
        key = key.strip
        value = value.strip

        # Strip quotes
        if (value.start_with?('"') && value.end_with?('"')) ||
           (value.start_with?("'") && value.end_with?("'"))
          value = value[1..-2]
        end

        # Strip prefix if provided
        if prefix && key.start_with?(prefix)
          key = key[prefix.length..]
          key = key.delete_prefix("_") # remove leading underscore after prefix
        end

        data[key.downcase] = Coercion.coerce(value)
      end

      data
    end

    class << self
      private

      # Detect packages in a monorepo based on type.
      #
      # @param root [String] monorepo root
      # @param type [Symbol] monorepo type
      # @param marker_path [String] path to the marker file
      # @return [Array<String>]
      def detect_packages(root, type, marker_path)
        case type
        when :pnpm
          parse_pnpm_workspace(marker_path, root)
        when :lerna
          parse_lerna_config(marker_path, root)
        when :turbo, :nx
          # Turbo/NX typically use package.json workspaces
          pkg_json = File.join(root, "package.json")
          if File.exist?(pkg_json)
            begin
              pkg = JSON.parse(File.read(pkg_json))
              return resolve_workspace_globs(root, pkg["workspaces"]) if pkg.key?("workspaces")
            rescue JSON::ParserError
              # ignore
            end
          end
          []
        else
          []
        end
      end

      # Parse pnpm-workspace.yaml for package globs.
      #
      # @param path [String] path to pnpm-workspace.yaml
      # @param root [String] monorepo root
      # @return [Array<String>]
      def parse_pnpm_workspace(path, root)
        # Simple YAML parsing without requiring the yaml gem for this
        globs = []
        in_packages = false

        File.readlines(path, chomp: true).each do |line|
          if line.match?(/\Apackages\s*:/)
            in_packages = true
            next
          elsif line.match?(/\A\S/)
            in_packages = false
            next
          end

          if in_packages && (m = line.match(/\A\s*-\s*['"]?(.+?)['"]?\s*\z/))
            globs << m[1]
          end
        end

        resolve_workspace_globs(root, globs)
      end

      # Parse lerna.json for package directories.
      #
      # @param path [String] path to lerna.json
      # @param root [String] monorepo root
      # @return [Array<String>]
      def parse_lerna_config(path, root)
        config = JSON.parse(File.read(path))
        globs = config["packages"] || ["packages/*"]
        resolve_workspace_globs(root, globs)
      rescue JSON::ParserError
        []
      end

      # Resolve workspace glob patterns to actual directories.
      #
      # @param root [String] monorepo root
      # @param patterns [Array<String>, Hash, nil] glob patterns or workspaces config
      # @return [Array<String>]
      def resolve_workspace_globs(root, patterns)
        patterns = patterns["packages"] if patterns.is_a?(Hash) && patterns.key?("packages")
        return [] unless patterns.is_a?(Array)

        dirs = []
        patterns.each do |pattern|
          full_pattern = File.join(root, pattern)
          Dir.glob(full_pattern).each do |match|
            dirs << match if File.directory?(match)
          end
        end

        dirs.sort
      end
    end
  end
end

File: langs/python/src/dotlyte/workspace.py
"""Monorepo / workspace support for DOTLYTE.

Detects monorepo root (pnpm-workspace.yaml, turbo.json, nx.json, lerna.json),
loads root-level .env first, then package-level .env on top.

Example::

    from dotlyte.workspace import load_workspace

    config = load_workspace(packages=["apps/web"])
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Literal

from dotlyte.loader import load


@dataclass
class MonorepoInfo:
    """Information about a detected monorepo."""

    root: str
    type: Literal["pnpm", "npm", "yarn", "nx", "turbo", "lerna", "unknown"]
    packages: list[str] = field(default_factory=list)


def find_monorepo_root(cwd: str | Path | None = None) -> MonorepoInfo | None:
    """Auto-detect the monorepo root by walking up from *cwd*.

    Checks for ``pnpm-workspace.yaml``, ``turbo.json``, ``nx.json``,
    ``lerna.json``, or ``package.json`` with a ``workspaces`` field.

    Args:
        cwd: Starting directory. Defaults to the current working directory.

    Returns:
        A ``MonorepoInfo`` if found, else ``None``.

    """
    directory = Path(cwd).resolve() if cwd else Path.cwd()
    root = Path(directory.anchor)

    while directory != root:
        info = _detect_monorepo_at(directory)
        if info is not None:
            return info
        parent = directory.parent
        if parent == directory:
            break
        directory = parent

    return None


def _detect_monorepo_at(directory: Path) -> MonorepoInfo | None:
    """Check *directory* for monorepo markers."""
    # pnpm workspaces
    pnpm_ws = directory / "pnpm-workspace.yaml"
    if pnpm_ws.is_file():
        return MonorepoInfo(
            root=str(directory),
            type="pnpm",
            packages=_extract_pnpm_workspaces(pnpm_ws),
        )

    # Turbo
    turbo_json = directory / "turbo.json"
    if turbo_json.is_file():
        return MonorepoInfo(
            root=str(directory),
            type="turbo",
            packages=_extract_package_json_workspaces(directory),
        )

    # Nx
    nx_json = directory / "nx.json"
    if nx_json.is_file():
        return MonorepoInfo(
            root=str(directory),
            type="nx",
            packages=_extract_package_json_workspaces(directory),
        )

    # Lerna
    lerna_json = directory / "lerna.json"
    if lerna_json.is_file():
        return MonorepoInfo(
            root=str(directory),
            type="lerna",
            packages=_extract_lerna_packages(lerna_json),
        )

    # npm / yarn workspaces
    pkg_path = directory / "package.json"
    if pkg_path.is_file():
        try:
            pkg = json.loads(pkg_path.read_text(encoding="utf-8"))
            if "workspaces" in pkg:
                is_yarn = (directory / "yarn.lock").is_file()
                return MonorepoInfo(
                    root=str(directory),
                    type="yarn" if is_yarn else "npm",
                    packages=_extract_from_workspaces_field(pkg["workspaces"]),
                )
        except (json.JSONDecodeError, OSError):
            pass

    return None


def get_shared_env(
    root: str | Path,
    prefix: str | None = None,
) -> dict[str, str]:
    """Load shared env keys from the monorepo root ``.env``.

    Args:
        root: Monorepo root directory.
        prefix: Optional prefix to strip from keys.

    Returns:
        Dictionary of shared key-value pairs.

    """
    env_path = Path(root) / ".env"
    if not env_path.is_file():
        return {}

    result: dict[str, str] = {}
    content = env_path.read_text(encoding="utf-8")

    for line in content.splitlines():
        trimmed = line.strip()
        if not trimmed or trimmed.startswith("#"):
            continue
        eq_idx = trimmed.find("=")
        if eq_idx < 0:
            continue

        key = trimmed[:eq_idx].strip()
        value = trimmed[eq_idx + 1 :].strip()

        if prefix and key.startswith(prefix):
            key = key[len(prefix) :]

        result[key] = value

    return result


def load_workspace(
    root: str | Path | None = None,
    packages: list[str] | None = None,
    shared_env_file: str | None = None,
    prefix: str | None = None,
    **load_options: Any,
) -> Any:
    """Load configuration with monorepo-aware env inheritance.

    Priority (highest wins):
        1. Environment variables
        2. Package ``.env`` files
        3. Root ``.env`` files
        4. Defaults

    Args:
        root: Monorepo root (auto-detected if omitted).
        packages: Sub-package paths relative to root (e.g. ``["apps/web"]``).
        shared_env_file: Explicit shared env file path at root.
        prefix: Env var prefix to strip.
        **load_options: Forwarded to ``load()``.

    Returns:
        A ``Config`` object.

    """
    mono_root = Path(root) if root else None

    if mono_root is None:
        info = find_monorepo_root()
        if info is not None:
            mono_root = Path(info.root)

    if mono_root is None:
        # Not a monorepo — fall back to regular load
        return load(prefix=prefix, **load_options)

    # Collect files: root .env first (lower priority), then package .env
    env_name = load_options.get("env")
    root_files = _resolve_env_files(mono_root, env_name)

    if shared_env_file:
        shared_path = mono_root / shared_env_file
        if shared_path.is_file():
            root_files.insert(0, str(shared_path))

    package_files: list[str] = []
    if packages:
        for pkg in packages:
            pkg_dir = mono_root / pkg
            package_files.extend(_resolve_env_files(pkg_dir, env_name))

    all_files = root_files + package_files
    existing_files = [f for f in all_files if Path(f).is_file()]

    return load(
        files=existing_files if existing_files else None,
        prefix=prefix,
        **load_options,
    )


# ──── Internal helpers ────


def _resolve_env_files(directory: Path, env: str | None = None) -> list[str]:
    """Return a list of potential .env file paths in *directory*."""
    files = [
        str(directory / ".env"),
        str(directory / ".env.local"),
    ]
    if env:
        files.append(str(directory / f".env.{env}"))
        files.append(str(directory / f".env.{env}.local"))
    return files


def _extract_pnpm_workspaces(filepath: Path) -> list[str]:
    """Parse ``pnpm-workspace.yaml`` for package globs."""
    try:
        content = filepath.read_text(encoding="utf-8")
        packages: list[str] = []
        in_packages = False

        for line in content.splitlines():
            trimmed = line.strip()
            if trimmed == "packages:":
                in_packages = True
                continue
            if in_packages and trimmed.startswith("- "):
                packages.append(trimmed[2:].strip("'\" "))
            elif in_packages and not trimmed.startswith("-") and trimmed:
                break

        return packages
    except OSError:
        return []


def _extract_package_json_workspaces(directory: Path) -> list[str]:
    """Read ``workspaces`` from ``package.json``."""
    pkg_path = directory / "package.json"
    if not pkg_path.is_file():
        return []
    try:
        pkg = json.loads(pkg_path.read_text(encoding="utf-8"))
        return _extract_from_workspaces_field(pkg.get("workspaces"))
    except (json.JSONDecodeError, OSError):
        return []


def _extract_from_workspaces_field(workspaces: Any) -> list[str]:
    """Normalise a ``workspaces`` field to a list of strings."""
    if isinstance(workspaces, list):
        return [w for w in workspaces if isinstance(w, str)]
    if isinstance(workspaces, dict) and "packages" in workspaces:
        pkgs = workspaces["packages"]
        if isinstance(pkgs, list):
            return [w for w in pkgs if isinstance(w, str)]
    return []


def _extract_lerna_packages(filepath: Path) -> list[str]:
    """Parse ``lerna.json`` for packages."""
    try:
        content = json.loads(filepath.read_text(encoding="utf-8"))
        pkgs = content.get("packages")
        if isinstance(pkgs, list):
            return [w for w in pkgs if isinstance(w, str)]
        return ["packages/*"]
    except (json.JSONDecodeError, OSError):
        return []
